### PR TITLE
Feature/device selectors

### PR DESCRIPTION
**Pull request by diptorupd**
This PR is extracted out of #271 and only adds the device selector classes to C API and Python API. 

- A `SyclDevice` can now be created out of SYCL's standard device selector classes. 
- A `SyclDevice` can also be constructed using a filter selector string.

### CODE REVIEW

File: dpctl/_sycl_device.pyx [hunk 1]
```diff
@@ -20,11 +20,41 @@
 """ Implements SyclDevice Cython extension type.
 """
 
-from ._backend cimport *
+from ._backend cimport (
```
Comment (oleksandr-pavlyk) [inline dpctl/_sycl_device.pyx:23]: So much better :)

**Review by oleksandr-pavlyk — APPROVED** [review]

**Comment by oleksandr-pavlyk** [comment]
Looking good. I still would like to build it locally, so please wait some before merging.

**Comment by diptorupd** [comment]
> Looking good. I still would like to build it locally, so please wait some before merging.

Sure, I will also like to add the support for filter selector. I tried the following code: It works, but breaks the `SyclDevice._create` implementation. How to design support for filter selector?

```python=
    @staticmethod
    cdef _init_helper(SyclDevice device, DPCTLSyclDeviceRef DRef):
        device._device_ref = DRef
        device._vendor_name = DPCTLDevice_GetVendorName(DRef)
        device._device_name = DPCTLDevice_GetName(DRef)
        device._driver_version = DPCTLDevice_GetDriverInfo(DRef)
        device._max_compute_units = DPCTLDevice_GetMaxComputeUnits(DRef)
        device._max_work_item_dims = DPCTLDevice_GetMaxWorkItemDims(DRef)
        device._max_work_item_sizes = DPCTLDevice_GetMaxWorkItemSizes(DRef)
        device._max_work_group_size = DPCTLDevice_GetMaxWorkGroupSize(DRef)
        device._max_num_sub_groups = DPCTLDevice_GetMaxNumSubGroups(DRef)
        device._int64_base_atomics = DPCTLDevice_HasInt64BaseAtomics(DRef)
        device._int64_extended_atomics = (
            DPCTLDevice_HasInt64ExtendedAtomics(DRef)
        )

    def __cinit__(self, filter_str):
        cdef const char *filter_c_str = NULL
        if type(filter_str) is unicode:
            string = bytes(<unicode>filter_str, "utf-8")
            filter_c_str = string
        elif isinstance(filter_str, unicode):
            string = bytes(unicode(filter_str), "utf-8")
            filter_c_str = <unicode>string
        cdef DPCTLSyclDeviceSelectorRef DSRef = (
            DPCTLFilterSelector_Create(filter_c_str)
        )
        cdef DPCTLSyclDeviceRef DRef = DPCTLDevice_CreateFromSelector(DSRef)
        if DRef is NULL:
            raise ValueError("Device could not be created from provided filter")
        # Initialize the attributes of the SyclDevice object
        SyclDevice._init_helper(self, DRef)
        # Free up the device selector
        DPCTLDeviceSelector_Delete(DSRef)
```

**Comment by diptorupd** [comment]
@oleksandr-pavlyk this one is good to go. The CI chokes up due to `pytest` not being added to the recipe. I have opened another PR #276 that should fix the CI.

**Comment by diptorupd** [comment]
@oleksandr-pavlyk We actually do need the `SyclDevice._create` method as we are using it to create a `SyclDevice` from a `cdef DPCTLSyclDeviceRef`.

**Comment by diptorupd** [comment]
@oleksandr-pavlyk are we good to go with this?

**Comment by diptorupd** [comment]
merging as @oleksandr-pavlyk gave his blessings :)

**Pull request merged**